test/test_state.c

```diff
@@ -14,4 +19,65 @@ Test(test_state, test_random_state) {
     cr_assert(eq(int, HAVE_AMOUNT, state.pieces_have.num_elements));
     cr_assert(eq(int, WANT_AMOUNT, state.pieces_want.num_elements));
 }
+
+Test(test_state, test_peer_exhange) {
+    remove("/tmp/test_peer_exchange");
+    int fd_in = open("/tmp/test_peer_exchange", O_CREAT | O_APPEND | O_RDWR, S_IRUSR | S_IWUSR);    
+    if(fd_in < 0) {
+        puts("failed to open test_state fd");
+    }
+
+    // say we know about 4 random ports
+    client_state state = new_state();
+    add_port(&state, 1);
+    add_port(&state, 2);
+    add_port(&state, 3);
+    add_port(&state, 4);
+
+    // say we are connected to a "client" mocked as stdin
+    add_file_descriptor(&state, fd_in);
+
+    // attempt to broadcast to fds
+    peer_exchange(&state);
+    close(fd_in);
+
```

Comment:
I absolutely love how we can just pass any file descriptor as a "peer" and use it for unit tests